gradle-plugin/plugin/src/main/kotlin/com/emergetools/android/gradle/EmergePlugin.kt

```diff
@@ -470,6 +426,62 @@ class EmergePlugin : Plugin<Project> {
     }
   }
 
+  private fun registerReaperTasks(
+    appProject: Project,
+    extension: EmergePluginExtension,
+    variant: Variant,
+  ) {
+    registerReaperPreflightTask(appProject, extension, variant)
+    // Only register upload task if Reaper is enabled
+    if (extension.reaperOptions.enabled.getOrElse(false)) {
+      registerReaperUploadTask(appProject, extension, variant)
+    }
+  }
+
+  private fun registerReaperPreflightTask(
+    appProject: Project,
+    extension: EmergePluginExtension,
+    variant: Variant,
+  ) {
+    val preflightTaskName = "${EMERGE_TASK_PREFIX}ValidateReaper${variant.name.capitalize()}"
+    appProject.tasks.register(preflightTaskName, PreflightReaper::class.java) {
+      it.group = EMERGE_TASK_GROUP
+      it.description = "Validate Reaper is initialized for variant ${variant.name}"
+      it.reaperEnabled.set(extension.reaperOptions.enabled)
+      it.reaperPublishableApiKey.set(extension.reaperOptions.publishableApiKey)
+      it.mergedManifestFile.set(variant.artifacts.get(SingleArtifact.MERGED_MANIFEST))
+
+      // We want preflight to happen pretty early so we can detect error conditions and give them
+      // nice messages. Specifically we need it to occur prior to 'linking' steps which we detect
+      // that the Reaper added instrumentation calls methods in the SDK to avoid confusing error
+      // messages.
+      val minifyTaskName = "minify${variant.name.capitalize()}WithR8"
+      val minifyTasks = appProject.getTasksByName(minifyTaskName, false)
+      if (minifyTasks.size != 0) {
+        minifyTasks.forEach { minifyTask -> minifyTask.dependsOn(it) }
+      }
+    }
+  }
+
+  private fun registerReaperUploadTask(
+    appProject: Project,
+    extension: EmergePluginExtension,
+    variant: Variant,
+  ) {
+    val uploadReaperAabTaskName = "${EMERGE_TASK_PREFIX}UploadReaperAab${variant.name.capitalize()}"
+    val uploadReaperAabTask =
+      appProject.tasks.register(uploadReaperAabTaskName, InitializeReaper::class.java) {
+        it.artifact.set(variant.artifacts.get(SingleArtifact.BUNDLE))
+        it.setUploadTaskInputs(extension, appProject, variant)
+        it.setTagFromProductOptions(extension.reaperOptions, variant)
+      }
+    // Hook the bundle tasks to run the reaper upload task after they complete.
+    appProject.afterEvaluate { project ->
+      val bundleTask = project.tasks.named("bundle${variant.name.capitalize()}")
+      bundleTask.configure { it.finalizedBy(uploadReaperAabTask) }
```

Comment:
This sounds great but does it mean you can't build without an internet connection, does the Sentry plugin fail if you attempt to `bundleRelease` but don't manage to upload the bundle? I guess this is fine if Sentry does the same.

Reply:
(lgtm other than that)

Reply:
Only if `reaper.enabled` is set. Can always disable that (or remove since default is false), but I think this is ok. Alternatively we can add a `variants` property instead of enabled that would only hook the bundle task for the specified variants, but that is a bit of a fragile, magic-string based approach.

I would say we land this, get it in users' hands and gather feedback should that be an issue, and we can fast-follow with workarounds if it is.

Reply:
sg!